src/elements/datepicker/datepicker/datepicker.ts

```diff
@@ -393,110 +326,141 @@ export class SbbDatepickerElement extends LitElement {
     }
   }
 
-  /** Emits the change event. */
-  private _emitChange(date: Date): void {
-    this._setAriaLiveMessage(date);
+  private _emitInputUpdated(): void {
+    const { disabled, readOnly: readonly, min, max } = this._inputElement ?? {};
+    this._inputUpdated.emit({ disabled, readonly, min, max });
+  }
 
+  private _handleInputChange(): void {
+    if (this._tryApplyFormatToInput()) {
+      return;
+    }
+    this._validateDate();
+    this._setAriaLiveMessage();
     this._change.emit();
     this._didChange.emit();
+  }
 
-    if (this._inputElement) {
-      this._inputElement.dispatchEvent(new InputEvent('input', { bubbles: true, composed: true }));
-      this._inputElement.dispatchEvent(
-        new CustomEvent('change', { bubbles: true, composed: true }),
-      );
+  private _tryApplyFormatToInput(): boolean {
+    if (!this._inputElement) {
+      return false;
     }
-  }
 
-  private _getValidValue(value: string): string {
-    if (!value) {
-      return '';
+    const formattedDate = this.valueAsDate ? this._format(this.valueAsDate!) : '';
+    if (formattedDate && this._inputElement.value !== formattedDate) {
+      this._inputElement.value = formattedDate;
+      this._inputElement.dispatchEvent(new InputEvent('input', { bubbles: true, composed: true }));
+      this._inputElement.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
+      return true;
     }
 
-    const match: RegExpMatchArray | null = value.match(FORMAT_DATE);
+    return false;
+  }
 
-    if (match?.index === 0) {
-      return this._parseAndFormatValue(value);
-    } else if (Number.isInteger(+value)) {
-      return this._createAndComposeDate(+value);
-    } else if (this._dateAdapter.isValid(new Date(value))) {
-      return this._createAndComposeDate(value);
+  private _validateDate(): void {
+    if (!this._inputElement) {
+      return;
     }
 
-    return value;
+    const isEmptyOrValid =
+      !this._inputElement.value ||
+      (!!this.valueAsDate &&
+        isDateAvailable(
+          this.valueAsDate,
+          this.dateFilter,
+          this._inputElement?.min,
+          this._inputElement?.max,
+        ));
+    const wasValid = !this._inputElement.hasAttribute('data-sbb-invalid');
+    this._inputElement.toggleAttribute('data-sbb-invalid', !isEmptyOrValid);
+    if (wasValid !== isEmptyOrValid) {
+      this._validationChange.emit({ valid: isEmptyOrValid });
+    }
   }
 
-  private _parse(value: string): Date | undefined {
-    return this.dateParser ? this.dateParser(value) : this._dateAdapter.parse(value, this.now);
+  private _parseInput(deserializeAsFallback = false): void {
+    const value = this._inputElement?.value ?? '';
+    this._valueAsDate = this._dateAdapter.getValidDateOrNull(
+      this.dateParser
+        ? this.dateParser(value)
+        : this._dateAdapter.parse(value, this.now) ??
+            (deserializeAsFallback ? this._dateAdapter.deserialize(value) : null),
+    );
```

Comment:
Probably not the most readable bit of code 😅
Not wrong though

Reply:
That's fair. 😅 
I'll check how I can improve it a bit.